### ISSUE TITLE
internal/graphicsdriver/opengl,metal: Avoid uploading uniform variables when possible

### ISSUE THREAD

**Issue by hajimehoshi** [hajimehoshi/ebiten#1848]
/CC @Zyko0

**Comment by Zyko0** [comment]
I don't think this impacts performance on recent hardware : https://community.khronos.org/t/performance-hit-gluniform/61202/10

I couldn't make a difference happen by testing with the example in #1862 with 4 shaders, and a bunch of uniforms each (win10, rtx 2080):

3000~ fps
```
go run ./examples/issue_1848/main.go
```

3000~ fps
```
git checkout a3eb8933e5 ./internal/graphicsdriver/opengl/program.go
go run ./examples/issue_1848/main.go
```

**Comment by Zyko0** [comment]
As a conclusion, caching by program seems correct on opengl (since there's already some sort of caching on Ebiten's internal).
We did prove that there's no relevant performance impact on different specs and OSes, so there's nothing immediately beneficial for implementing this.

Maybe the `performance` tag can go, for another one like `fix` or something

**Comment by hajimehoshi** [comment]
> edit: also I don't know if opengl is a tag or if there's metal to remove from the title maybe

There is `ebitengl` tag to force to use GL.

**Comment by hajimehoshi** [comment]
Let me deprioritize this task, but keep this issue open. When we find new things, let's revisit this.

**Comment by Zyko0** [comment]
> There is ebitengl tag to force to use GL.

I meant a github tag* on this issue sorry
